Libraries/LibWeb/CredentialManagement/PasswordCredentialOperations.cpp

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright (c) 2025, Kenneth Myhra <kennethmyhra@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <LibWeb/CredentialManagement/PasswordCredentialOperations.h>
+#include <LibWeb/HTML/HTMLFormElement.h>
+#include <LibWeb/XHR/FormData.h>
+
+namespace Web::CredentialManagement {
+
+// https://www.w3.org/TR/credential-management-1/#abstract-opdef-create-a-passwordcredential-from-an-htmlformelement
+WebIDL::ExceptionOr<GC::Ref<PasswordCredential>> create_password_credential(JS::Realm& realm, GC::Ptr<HTML::HTMLFormElement> const form, URL::Origin const& origin)
+{
+    // 1. Let data be a new PasswordCredentialData dictionary.
+    PasswordCredentialData data;
+
+    // 2. Set data’s origin member’s value to origin’s value.
+    data.origin = origin.serialize();
+
+    // 3. Let formData be the result of executing the FormData constructor on form.
+    auto form_data = TRY(XHR::FormData::construct_impl(realm, form));
+
+    // 4. Let elements be a list of all the submittable elements whose form owner is form, in tree order.
+    auto elements = form->get_submittable_elements();
+
+    // 5. Let newPasswordObserved be false.
+    bool new_password_observed = false;
+
+    // 6. For each field in elements, run the following steps:
+    for (auto const& field : elements) {
+        dbgln("field->name(): {}", field->name());
+        // 1. If field does not have an autocomplete attribute, then skip to the next field.
+        if (auto attr = field->attribute(HTML::AttributeNames::autocomplete); !attr.has_value() || attr->is_empty())
+            continue;
+
+        // 2. Let name be the value of field’s name attribute.
+        // 3. If formData’s has() method returns false when executed on name, then skip to the next field.
+        auto name = field->attribute(HTML::AttributeNames::name);
+        if (!name.has_value() || !form_data->has(name.value()))
+            continue;
+
+        // 4. If field’s autocomplete attribute’s value contains one or more autofill detail tokens (tokens), then:
+        if (auto tokens = field->attribute(HTML::AttributeNames::autocomplete); tokens.has_value()) {
```

Comment:
suggestion: This is always true because the field is skipped by step 1 if there is no autocomplete attribute. It would be nice to have the check for `one or more autofill detail tokens`, but it's non-blocking since the following code works on a subset of known good tokens.

Reply:
Removed the if-statement and left step 4 as a FIXME for now.